File: application/external/metrics/metrics_service.py
from abc import ABC, abstractmethod

from application.dto.metrics.admin_metrics import AdminMetrics
from application.dto.metrics.app_metrics import AppMetrics
from application.dto.metrics.contractee_metrics import ContracteeMetrics
from application.dto.metrics.contractor_metrics import ContractorMetrics
from domain.dto.user.internal.user_context_dto import UserContextDTO


class MetricsService(ABC):
    @abstractmethod
    async def get_app_metrics(self) -> AppMetrics: ...
    @abstractmethod
    async def get_admin_metrics(
        self, context: UserContextDTO
    ) -> AdminMetrics: ...
    @abstractmethod
    async def get_contractee_metrics(
        self, context: UserContextDTO
    ) -> ContracteeMetrics: ...
    @abstractmethod
    async def get_contractor_metrics(
        self, context: UserContextDTO
    ) -> ContractorMetrics: ...

File: nginx/frontend/js/dao/MetricsService.js
import {
    AdminMetrics,
    AppMetrics,
    ContracteeMetrics,
    ContractorMetrics,
} from "../entities/Metrics.js";
import { ADMIN_BASE, CONTRACTEE_BASE, CONTRACTOR_BASE } from "./HttpClient.js";

const METRICS_APP_BASE = "metrics";
export const METRICS_BASE = {
    app: METRICS_APP_BASE,
    admin: `${ADMIN_BASE}/${METRICS_APP_BASE}`,
    contractee: `${CONTRACTEE_BASE}/${METRICS_APP_BASE}`,
    contractor: `${CONTRACTOR_BASE}/${METRICS_APP_BASE}`,
};

export class MetricsService {
    constructor(httpClient) {
        this.httpClient = httpClient;
    }

    async getAppMetrics() {
        const data = await this.httpClient.request(`/${METRICS_BASE.app}`);
        return new AppMetrics(data);
    }

    async getAdminMetrics() {
        const data = await this.httpClient.request(`/${METRICS_BASE.admin}`);
        return new AdminMetrics(data);
    }

    async getContracteeMetrics() {
        const data = await this.httpClient.request(
            `/${METRICS_BASE.contractee}`
        );
        return new ContracteeMetrics(data);
    }

    async getContractorMetrics() {
        const data = await this.httpClient.request(
            `/${METRICS_BASE.contractor}`
        );
        return new ContractorMetrics(data);
    }
}
